### PR TITLE
Extend IPC protocol by error channel

### PR DESCRIPTION
**Pull request by t-wissmann**
This makes the ipc protocol and herbstclient support the error channel.
In order to make herbstclient compatible with old herbstluftwm ipc
server implementations, the error channel support is indicated by an
atom on hlwm's ipc server window (which is so far only used for the hook
system). Conversely, an old herbstclient version should also work more
or less with the new herbstluftwm ipc server: it simply ignores the
error channel.

Due to this change, a few more missing `output.perror()` calls surfaced,
affecting implementation and test cases.

In order to test the new herbstclient, the test suite features a
minimalist ipc server implementation which can be used to simulate wrong
behaviour and thus to stress the error handling in the client.

### CODE REVIEW

**Comment by t-wissmann** [comment]
I think, now it is ready :)

File: tests/test_rules.py [hunk 1]
```diff
@@ -140,10 +142,11 @@ def test_rule_parse_error_printed_at_client(hlwm, command, rulearg, errormsg):
     else:
         full_cmd += [rulearg]
 
-    proc = hlwm.call(full_cmd)
-    # since the command does not fail, the error
+    proc = hlwm.unchecked_call(full_cmd)
+
+    # the command does not fail, but the error
     # message appears on stdout
```
Comment (blarz) [inline tests/test_rules.py:148]: stderr?

File: tests/test_herbstclient.py [hunk 2]
```diff
@@ -220,3 +225,175 @@ def test_ensure_newline(hlwm):
                            universal_newlines=True)
     assert proc1.stdout == "test\n"
     assert proc2.stdout == "test\n"
+
+
+class IpcServer:
+    """A simply re-implementation of the ipc server
```
Comment (blarz) [inline tests/test_herbstclient.py:231]: simple

File: src/ipc-protocol.h [hunk 3]
```diff
@@ -5,9 +5,17 @@
 //#define HERBST_IPC_READY "HERBST_IPC_READY"
 //#define HERBST_IPC_ATOM  "_HERBST_IPC"
 #define HERBST_IPC_ARGS_ATOM "_HERBST_IPC_ARGS"
+//! the (utf8 text) atom containing the output channel
 #define HERBST_IPC_OUTPUT_ATOM "_HERBST_IPC_OUTPUT"
+//! the (utf8 text) atom containing the error channel
+#define HERBST_IPC_ERROR_ATOM "_HERBST_IPC_ERROR"
 #define HERBST_IPC_STATUS_ATOM "_HERBST_IPC_EXIT_STATUS"
 
+/** if this (integer) atom present and non-empty on the hook window,
```
Comment (blarz) [inline src/ipc-protocol.h:14]: "is" missing?

File: ipc-client/ipc-client.h [hunk 4]
```diff
@@ -23,8 +23,19 @@ void hc_disconnect(HCConnection* con);
...
+/**
+ * @brief send a command and wait for its reply to arrive
+ * @param the connection to hlwm
+ * @param number of arguments of the ipc call
+ * @param the arguments of the ipc call
+ * @param the returned output channel (needs to be free'd by the caller on success)
+ * @param the returned error channel (needs to be free'd by the caller on success)
+ * @param the returned exit status
+ * @return whether the reply was received successfully.
+ * if false is returned, then none of the ret_⎄-pointers is set.
```
Comment (blarz) [inline ipc-client/ipc-client.h:35]: ret_⎄ strange unicode garbage?

**Review by blarz — CHANGES_REQUESTED** [review]
Minor findings

**Comment by blarz** [comment]
Not related to this change but maybe for another PR:
```
out << "error: Command \"" << args.command() << "\" not found" << endl;
```
in command.cpp could also be moved to stderr somehow?

**Comment by t-wissmann** [comment]
Thanks for looking through the long diff :-)
I've fixed all the points you mentioned and created a new PR for the "command not found" error: #1286

**Review by blarz — APPROVED** [review]
Looks good to me, played around with it locally, too

**Pull request merged**